crates/edr_evm/src/spec.rs

```diff
@@ -512,3 +462,130 @@ impl RuntimeSpec for L1ChainSpec {
+        BlockEnv {
+            number: header.number,
+            beneficiary: header.beneficiary,
+            timestamp: header.timestamp,
+            difficulty: header.difficulty,
+            basefee: header.base_fee_per_gas.map_or(0u64, |base_fee| {
+                base_fee.try_into().expect("base fee is too large")
+            }),
+            gas_limit: header.gas_limit,
+            prevrandao: if hardfork >= l1::SpecId::MERGE {
+                Some(header.mix_hash)
+            } else {
+                None
+            },
+            blob_excess_gas_and_price: header.blob_gas.as_ref().map(
+                |BlobGas { excess_gas, .. }| {
+                    eip4844::BlobExcessGasAndPrice::new(*excess_gas, hardfork >= l1::SpecId::PRAGUE)
+                },
+            ),
+        }
+    }
+}
+
+#[cfg(test)]
+mod l1_chain_spec_tests {
+
+    use edr_eth::{
+        block::{BlobGas, Header},
+        l1, Address, Bloom, Bytes, B256, B64, U256,
+    };
+
+    use crate::spec::{BlockEnvConstructor as _, L1ChainSpec};
+
+    fn build_block_header(blob_gas: Option<BlobGas>) -> Header {
+        Header {
+            parent_hash: B256::default(),
+            ommers_hash: B256::default(),
+            beneficiary: Address::default(),
+            state_root: B256::default(),
+            transactions_root: B256::default(),
+            receipts_root: B256::default(),
+            logs_bloom: Bloom::default(),
+            difficulty: U256::default(),
+            number: 124,
+            gas_limit: u64::default(),
+            gas_used: 1337,
+            timestamp: 0,
+            extra_data: Bytes::default(),
+            mix_hash: B256::default(),
+            nonce: B64::from(99u64),
+            base_fee_per_gas: None,
+            withdrawals_root: None,
+            blob_gas,
+            parent_beacon_block_root: None,
+            requests_hash: Some(B256::random()),
+        }
+    }
+
+    #[test]
+    fn generic_block_constructor_should_not_default_excess_blob_gas_for_cancun() {
+        let header = build_block_header(None); // No blob gas information
+
+        let block = L1ChainSpec::new_block_env(&header, l1::SpecId::CANCUN);
+        assert_eq!(block.blob_excess_gas_and_price, None);
+    }
+
+    #[test]
+    fn generic_block_constructor_should_not_default_excess_blob_gas_below_cancun() {
+        let header = build_block_header(None); // No blob gas information
+
+        let block = L1ChainSpec::new_block_env(&header, l1::SpecId::SHANGHAI);
+        assert_eq!(block.blob_excess_gas_and_price, None);
+    }
+
+    #[test]
+    fn generic_block_constructor_should_not_default_excess_blob_gas_above_cancun() {
+        let header = build_block_header(None); // No blob gas information
+
+        let block = L1ChainSpec::new_block_env(&header, l1::SpecId::PRAGUE);
+        assert_eq!(block.blob_excess_gas_and_price, None);
+    }
+
+    #[test]
+    fn generic_block_constructor_should_use_existing_excess_blob_gas() {
```

Comment:
```suggestion
    fn l1_block_constructor_should_not_default_excess_blob_gas_for_cancun() {
        let header = build_block_header(None); // No blob gas information

        let block = L1ChainSpec::new_block_env(&header, l1::SpecId::CANCUN);
        assert_eq!(block.blob_excess_gas_and_price, None);
    }

    #[test]
    fn l1_block_constructor_should_not_default_excess_blob_gas_before_cancun() {
        let header = build_block_header(None); // No blob gas information

        let block = L1ChainSpec::new_block_env(&header, l1::SpecId::SHANGHAI);
        assert_eq!(block.blob_excess_gas_and_price, None);
    }

    #[test]
    fn l1_block_constructor_should_not_default_excess_blob_gas_after_cancun() {
        let header = build_block_header(None); // No blob gas information

        let block = L1ChainSpec::new_block_env(&header, l1::SpecId::PRAGUE);
        assert_eq!(block.blob_excess_gas_and_price, None);
    }

    #[test]
    fn l1_block_constructor_should_use_existing_excess_blob_gas() {
```

Should these all be named `l1_*`?

Reply:
indeed, thanks for the catch 🙇